### DISCUSSION TITLE
Resource Not Available in Later System Call

### DISCUSSION THREAD

**Discussion by asb9189** [bevyengine/bevy#8276] [Q&A]
Hi All,

I have a ```setup``` system that adds a resource containing the primary window's width and height so that I don't need to query for it every time. In other words, I want my future systems to have ```Res<WinSize>``` rather than ```Query<&Window, With<PrimaryWindow>>``` in their function definition.

However, my ```spawn_player``` system runs after ```setup``` and panics because the ```Res<WinSize``` has not yet been added as an available resource. The ordering of system execution is correct, however, the commands have not yet been executed by the time ```spawn_player``` executes.

I found this useful discussion (https://github.com/bevyengine/bevy/discussions/3174) regarding the same issue, however it's not for Bevy 0.10. I noticed that Bevy 0.10's migration notes says the following:

```
Stages have been removed. Replace these with system sets, and then add command flushes using the apply_system_buffers exclusive system where needed.
```

I'm new to the Bevy ecosystem so I've been having a hard time finding examples of ```system sets``` and using the ```apply_system_buffers``` function. How could I easily ensure a resource that is injected by one system is available by a later system?

Thanks!

**Comment by nicopap** [comment]
How are you adding `Res<WinSize>` to your app? If you add `setup` to your app using `.add_system(Startup, setup)` and `spawn_player` in `Update` afaik, it should work, there is a synchronization step between the two. So it's something else.

Let me suggest two alternatives

1. Instead of adding the resource in `setup`, implement `FromWorld` for it.
```rust
impl FromWorld for WinSize {
   fn from_world(world: &mut World) -> Self {
      // impl
   }
}
```

Then you need to add the resource to the app as follow: `app.init_resource::<WinSize>()`.

2. Use `Query<&Window, With<PrimaryWindow>>`. After all, your resource would lie to you if the window size changes from external factors. `WinSize` doesn't follow the single-source-of-truth principle, and will likely lead to bugs.

To simplify the task, you can define a `SystemParam` as follow:
```rust
#[derive(SystemParam)]
struct WinSize<'w, 's> {
    windows: Query<'w, 's, &Window, With<PrimaryWindow>>,
}
impl<'w, 's> WinSize<'w, 's> {
    fn size(&self) -> (u32, u32) {
        let resolution = self.windows.get_single().unwrap().resolution;
        (resolution.width, resolution.height)
    }
}
```

Now if you want to query for window size in a system you do
```rust
fn win_size_system(win_size: WinSize) {
  let size = win_size.size();
  println!("{size:?}");
}
```
